internal/trie/node/encode.go

```diff
@@ -0,0 +1,75 @@
+// Copyright 2021 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package node
+
+import (
+	"fmt"
+
+	"github.com/ChainSafe/gossamer/internal/trie/codec"
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/pkg/scale"
+)
+
+// Encode encodes the node to the buffer given.
+// The encoding format is documented in encode_doc.go.
+func (n *Node) Encode(buffer Buffer) (err error) {
+	if !n.Dirty && n.Encoding != nil {
+		_, err = buffer.Write(n.Encoding)
+		if err != nil {
+			return fmt.Errorf("cannot write stored encoding to buffer: %w", err)
+		}
+		return nil
+	}
+
+	err = encodeHeader(n, buffer)
+	if err != nil {
+		return fmt.Errorf("cannot encode header: %w", err)
+	}
+
+	keyLE := codec.NibblesToKeyLE(n.Key)
+	_, err = buffer.Write(keyLE)
+	if err != nil {
+		return fmt.Errorf("cannot write LE key to buffer: %w", err)
+	}
+
+	if n.Type() == Branch {
+		childrenBitmap := common.Uint16ToBytes(n.ChildrenBitmap())
+		_, err = buffer.Write(childrenBitmap)
+		if err != nil {
+			return fmt.Errorf("cannot write children bitmap to buffer: %w", err)
+		}
+	}
+
+	// check value is not nil for branch nodes, even though
+	// leaf nodes always have a non-nil value.
+	if n.Type() == Leaf || n.Value != nil {
+		// TODO remove `n.Type() == Leaf` and update tests
```

Comment:
Well, yes and no, since changing this will affect test values and I wanted to keep this PR with the same test values to be sure it doesn't change behavior. I'd rather do it right after it gets merged as another PR. Let's keep this unresolved though so I don't forget :wink:

Reply:
@EclesioMeloJunior done in #2588 resolving this!